x/emissions/module/rewards/rewards.go

```diff
@@ -45,26 +46,46 @@ func EmitRewards(
 	if uint64(len(sortedRewardableTopics)) > moduleParams.MaxActiveTopicsPerBlock {
 		sortedRewardableTopics = sortedRewardableTopics[:moduleParams.MaxActiveTopicsPerBlock]
 	}
-	// Get total weight of rewardable topics
-	sumWeightOfRewardableTopics := alloraMath.ZeroDec()
+
+	// Get the global total sum of previous topic weights
+	totalWeightOfActiveTopics, err := k.GetTotalSumPreviousTopicWeights(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get total sum of previous topic weights")
+	}
+	// Get epoch lengths for sorted rewardable topics
+	epochLengths := make(map[uint64]int64)
 	for _, topicId := range sortedRewardableTopics {
-		sumWeightOfRewardableTopics, err = sumWeightOfRewardableTopics.Add(*weights[topicId])
+		topic, err := k.GetTopic(ctx, topicId)
```

Comment:
During the `EndBlocker` we do `GetTopic` multiple times: 
- On `GetAndUpdateActiveTopicWeights`
- On `UpdateNoncesOfActiveTopics`
- On `EmitRewards` 
- On `getWorkersRewardFractions`
- On `pruneRecordsAfterRewards`

Probably we can optimize it... 

I know 80% of this wasn't introduced in your PR, so lmk if you think it's better to create a separate ticket to optimize this

Reply:
yeah probably wiser to separate that , to separate responsibiitlies. This PR has enough to worry about lol